third_party/com_github_aws_cpp_sdk/aws_cpp_sdk.BUILD.bazel

```diff
@@ -0,0 +1,180 @@
+# Description:
+#   AWS C++ SDK
+
+package(default_visibility = ["//visibility:public"])
+
+licenses(["notice"])  # Apache 2.0
+
+exports_files(["LICENSE"])
+
+cc_library(
+    name = "core",
+    srcs = glob([
+        "aws-cpp-sdk-core/source/*.cpp",  # AWS_SOURCE
+        "aws-cpp-sdk-core/source/external/tinyxml2/*.cpp",  # AWS_TINYXML2_SOURCE
+        "aws-cpp-sdk-core/source/external/cjson/*.cpp",  # CJSON_SOURCE
+        "aws-cpp-sdk-core/source/auth/*.cpp",  # AWS_AUTH_SOURCE
```

Comment:
Naively, I assume the challenge will be to see how much of the source can be excised, while retaining the auth capalities.

Reply:
We now have tinyxml2, so  it would be good to remove that if we can. 

:)

Reply:
Indeed, lets see whats possible here.

Reply:
It looks like the `tinyxml2` vendored into the AWS c++ sdk has been patched. Quoting from  https://github.com/aws/aws-sdk-cpp/blob/d79ece080eba11212b10a857ae551a4b8423fe00/src/aws-cpp-sdk-core/include/aws/core/external/tinyxml2/tinyxml2.h#L23-L28.

```
  (1) Memory management operations use aws memory management api
  (2) Import-export preprocessor logic tweaked for better integration into core library
  (3) Wrapped everything in Amazon namespace to prevent static linking issues if the user includes a version of this code through another dependency
```

Would it be reasonable to use the AWS C++ SDK version of tinyxml2, instead of the vanilla version currently used with tensorflow?

Reply:
> Naively, I assume the challenge will be to see how much of the source can be excised, while retaining the auth capalities.

Naively. So everything in the core library is tightly coupled. It doesn't seem possible to easily excise portions of the core library out. However, here are the `.so` binary sizes after stripping:

```bash
$ ~/code/tensorstore/bazel-bin/external$ ls -lh com_github_aws_cpp_sdk/libcore.so
-rwxr-xr-x 1 simon simon 3.8M Apr  3 15:42 com_github_aws_cpp_sdk/libcore.so

$ ls -lh com_github_aws_c_common/libaws_c_common.so
-rwxr-xr-x 1 simon simon 193K Apr  3 15:40 com_github_aws_c_common/libaws_c_common.so

$ ls -lh com_github_aws_c_event_stream/libaws_c_event_stream.so
-rwxr-xr-x 1 simon simon 35K Apr  3 15:40 com_github_aws_c_event_stream/libaws_c_event_stream.so

$ ls -lh com_github_aws_checksums/libaws_checksums.so
-rwxr-xr-x 1 simon simon 43K Apr  3 15:40 com_github_aws_checksums/libaws_checksums.so
```

Lets round that up and say that it costs `~4.1MB total` to get the AWS SDK code functionality. Considering that some of the functionality is already replicated already in S3RequestBuilder, depending on `core` may not be a bad tradeoff in terms of code maintenance.

Reply:
In most cases outside tests the AWS libraries would be statically linked to tensorstore and the linker may automatically remove most unused code anyway.

Reply:
> In most cases outside tests the AWS libraries would be statically linked to tensorstore and the linker may automatically remove most unused code anyway.

Yes, the `.so` sizes would only establish a rough upper bound on include size. I'm trying to estimate before making significant code changes, but I guess including the code and inspecting the final wheel (for e.g.) is the only way to be sure.

`libs3.so` is `9.2MB` stripped.